engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala

```diff
@@ -0,0 +1,97 @@
+package cromwell.engine.workflow.lifecycle.execution
+
+import akka.actor.{ActorRef, LoggingFSM, Props}
+import cromwell.backend.BackendJobExecutionActor._
+import cromwell.backend.{BackendInitializationData, BackendJobDescriptor, BackendJobDescriptorKey, BackendLifecycleActorFactory}
+import cromwell.core.logging.WorkflowLogging
+import cromwell.engine.workflow.lifecycle.execution.EngineJobExecutionActor._
+import cromwell.engine.workflow.lifecycle.execution.JobPreparationActor.{BackendJobPreparationFailed, BackendJobPreparationSucceeded}
+import cromwell.jobstore._
+
+object EngineJobExecutionActor {
+  /** States */
+  sealed trait EngineJobExecutionActorState
+  case object Pending extends EngineJobExecutionActorState
+  case object CheckingJobStatus extends EngineJobExecutionActorState
+  case object PreparingJob extends EngineJobExecutionActorState
+  case object RunningJob extends EngineJobExecutionActorState
+
+  /** Commands */
+  sealed trait EngineJobExecutionActorCommand
+  case class Start(jobKey: BackendJobDescriptorKey) extends EngineJobExecutionActorCommand
+  case class Restart(jobKey: BackendJobDescriptorKey) extends EngineJobExecutionActorCommand
+
+  case class JobRunning(jobDescriptor: BackendJobDescriptor, backendJobExecutionActor: ActorRef)
+
+  def props(executionData: WorkflowExecutionActorData, factory: BackendLifecycleActorFactory,
+            initializationData: Option[BackendInitializationData]) = {
+    Props(new EngineJobExecutionActor(executionData, factory, initializationData))
+  }
+
+  /** Data */
+  case class EngineJobExecutionActorData(restarting: Boolean = false)
+}
+
+class EngineJobExecutionActor(executionData: WorkflowExecutionActorData,
+                              factory: BackendLifecycleActorFactory,
+                              initializationData: Option[BackendInitializationData]) extends LoggingFSM[EngineJobExecutionActorState, EngineJobExecutionActorData] with WorkflowLogging {
+
+  override val workflowId = executionData.workflowDescriptor.id
+
+  startWith(Pending, EngineJobExecutionActorData())
+
+  when(Pending) {
+    case Event(Start(jobKey), _) =>
+      prepareJob(jobKey) using EngineJobExecutionActorData(restarting = false)
+    case Event(Restart(jobKey), _) =>
+      context.actorOf(JobStoreReader.props()) ! QueryJobCompletion(jobKey)
+      goto(CheckingJobStatus) using EngineJobExecutionActorData(restarting = true)
+  }
+
+  when(CheckingJobStatus) {
+    case Event(JobNotComplete(jobKey), _) =>
+      prepareJob(jobKey)
+    case Event(JobComplete(jobKey, jobResult), _) =>
+      jobResult match {
+        case JobResultSuccess(returnCode, jobOutputs) =>
+          context.parent ! new SucceededResponse(jobKey, returnCode, jobOutputs)
+          context stop self
+          stay()
+        case JobResultFailure(returnCode, reason) =>
+          context.parent ! new FailedNonRetryableResponse(jobKey, reason, returnCode)
```

Comment:
Currently the JobStore doesn't give any indication wether or not it is retryable.